### ISSUE TITLE
ParseError Message: syntax error, unexpected '?'

### ISSUE THREAD

**Issue by diplodata** [nextcloud/cms_pico#55]
I'm struggling to get this awesome looking app to work (Nextcloud 13, Ubuntu 16.04). I've installed as per instructions (as far as I can understand them - the apache stuff is a bit light), and set up a demo page called `welcome`.  Calling `https://mycloud/sites/welcome/?sub/welcome` and `https://mycloud/sites/welcome` both return `Internal Server Error`.  The error log reports:

```
{
  "reqId": "9HBvrcZtg5BmSKzXyovD",
  "level": 3,
  "time": "2018-04-19T17:03:58+01:00",
  "remoteAddr": "x.x.x.x",
  "user": "myuser",
  "app": "index",
  "method": "GET",
  "url": "\/index.php\/apps\/cms_pico\/pico\/welcome",
  "message": "Exception: {\"Exception\":\"ParseError\",\"Message\":\"syntax error, unexpected '?'\",\"Code\":0,\"Trace\":\"#0 \\\/var\\\/www\\\/nextcloud\\\/lib\\\/composer\\\/composer\\\/ClassLoader.php(321): Composer\\\\Autoload\\\\includeFile('\\\/var\\\/www\\\/nextcl...')\\n#1 [internal function]: Composer\\\\Autoload\\\\ClassLoader->loadClass('Symfony\\\\\\\\Compone...')\\n#2 \\\/var\\\/www\\\/nextcloud\\\/apps\\\/cms_pico\\\/vendor\\\/picocms\\\/pico\\\/lib\\\/Pico.php(785): spl_autoload_call('Symfony\\\\\\\\Compone...')\\n#3 \\\/var\\\/www\\\/nextcloud\\\/apps\\\/cms_pico\\\/lib\\\/Pico.php(81): Pico->parseFileMeta('---\\\\nTitle: Welc...', Array)\\n#4 \\\/var\\\/www\\\/nextcloud\\\/apps\\\/cms_pico\\\/vendor\\\/picocms\\\/pico\\\/lib\\\/Pico.php(321): OCA\\\\CMSPico\\\\Pico->parseFileMeta('---\\\\nTitle: Welc...', Array)\\n#5 \\\/var\\\/www\\\/nextcloud\\\/apps\\\/cms_pico\\\/lib\\\/Service\\\/PicoService.php(166): Pico->run()\\n#6 \\\/var\\\/www\\\/nextcloud\\\/apps\\\/cms_pico\\\/lib\\\/Service\\\/PicoService.php(109): OCA\\\\CMSPico\\\\Service\\\\PicoService->getContentFromPico(Object(OCA\\\\CMSPico\\\\Model\\\\Website))\\n#7 \\\/var\\\/www\\\/nextcloud\\\/apps\\\/cms_pico\\\/lib\\\/Service\\\/WebsitesService.php(249): OCA\\\\CMSPico\\\\Service\\\\PicoService->getContent(Object(OCA\\\\CMSPico\\\\Model\\\\Website))\\n#8 \\\/var\\\/www\\\/nextcloud\\\/apps\\\/cms_pico\\\/lib\\\/Controller\\\/PicoController.php(103): OCA\\\\CMSPico\\\\Service\\\\WebsitesService->getWebpageFromSite('welcome', 'myuser', '')\\n#9 \\\/var\\\/www\\\/nextcloud\\\/apps\\\/cms_pico\\\/lib\\\/Controller\\\/PicoController.php(88): OCA\\\\CMSPico\\\\Controller\\\\PicoController->getPage('welcome', '')\\n#10 [internal function]: OCA\\\\CMSPico\\\\Controller\\\\PicoController->getRoot('welcome')\\n#11 \\\/var\\\/www\\\/nextcloud\\\/lib\\\/private\\\/AppFramework\\\/Http\\\/Dispatcher.php(161): call_user_func_array(Array, Array)\\n#12 \\\/var\\\/www\\\/nextcloud\\\/lib\\\/private\\\/AppFramework\\\/Http\\\/Dispatcher.php(91): OC\\\\AppFramework\\\\Http\\\\Dispatcher->executeController(Object(OCA\\\\CMSPico\\\\Controller\\\\PicoController), 'getRoot')\\n#13 \\\/var\\\/www\\\/nextcloud\\\/lib\\\/private\\\/AppFramework\\\/App.php(115): OC\\\\AppFramework\\\\Http\\\\Dispatcher->dispatch(Object(OCA\\\\CMSPico\\\\Controller\\\\PicoController), 'getRoot')\\n#14 \\\/var\\\/www\\\/nextcloud\\\/lib\\\/private\\\/AppFramework\\\/Routing\\\/RouteActionHandler.php(47): OC\\\\AppFramework\\\\App::main('OCA\\\\\\\\CMSPico\\\\\\\\Con...', 'getRoot', Object(OC\\\\AppFramework\\\\DependencyInjection\\\\DIContainer), Array)\\n#15 [internal function]: OC\\\\AppFramework\\\\Routing\\\\RouteActionHandler->__invoke(Array)\\n#16 \\\/var\\\/www\\\/nextcloud\\\/lib\\\/private\\\/Route\\\/Router.php(297): call_user_func(Object(OC\\\\AppFramework\\\\Routing\\\\RouteActionHandler), Array)\\n#17 \\\/var\\\/www\\\/nextcloud\\\/lib\\\/base.php(998): OC\\\\Route\\\\Router->match('\\\/apps\\\/cms_pico\\\/...')\\n#18 \\\/var\\\/www\\\/nextcloud\\\/index.php(37): OC::handleRequest()\\n#19 {main}\",\"File\":\"\\\/var\\\/www\\\/nextcloud\\\/apps\\\/news\\\/vendor\\\/symfony\\\/yaml\\\/Parser.php\",\"Line\":510}",
  "userAgent": "Mozilla\/5.0 (Macintosh; Intel Mac OS X 10_13_4) AppleWebKit\/537.36 (KHTML, like Gecko) Chrome\/65.0.3325.181 Safari\/537.36",
  "version": "13.0.1.1"
}
```

From what I can tell the critical bit is `ParseError Message: syntax error, unexpected '?'`, and this relates to line 510 of `/var/www/nextcloud/apps/news/vendor/symfony/yaml/Parser.php`.  That line is:

```
    510     private function getNextEmbedBlock(int $indentation = null, bool $inSequence = false): ?string
    511     {
    512         $oldLineIndentation = $this->getCurrentLineIndentation();
    ...
```

Possibly the issue related to the `?string` at the end of that line?

I've been wondering if this is related to the version of PHP we're running.  I've been trying to upgrade to 7.2, so my /var/apache2/ contains:

```
lrwxrwxrwx 1 root root   29 Apr  6 08:03 php7.0.conf -> ../mods-available/php7.0.conf
lrwxrwxrwx 1 root root   29 Apr  6 08:03 php7.0.load -> ../mods-available/php7.0.load
lrwxrwxrwx 1 root root   29 Apr 19 14:56 php7.2.conf -> ../mods-available/php7.2.conf
lrwxrwxrwx 1 root root   29 Apr 19 14:56 php7.2.load -> ../mods-available/php7.2.load
```
However, when I run `a2dismod php7.0` and `a2enmod php7.2` Nextcloud completely breaks with `HTTP ERROR 500`.

Very grateful for assistance.

**Comment by diplodata** [comment]
Solution is upgrade to `PHP7.2`.  The following should get you some or all of the way:

```
# add-apt-repository ppa:ondrej/php # may be necessary
add-apt-repository ppa:ondrej/apache2

apt install php libapache2-mod-php
apt-get install php7.2-mysql php7.2-curl php7.2-json php7.2-cgi  php7.2 libapache2-mod-php7.2
apt-get install php7.2-redis
apt-get install php7.2-dom

a2dismod php7.0
a2enmod php7.2

update-alternatives --set php /usr/bin/php7.2

apt-get install php7.2-zip
a2enmod proxy_fcgi setenvif
a2enconf php7.2-fpm

systemctl restart apache2
```

**Comment by PhrozenByte** [comment]
Pico CMS for Nextcloud v1.0.0 is arriving and I'm currently working through some old issues that can be resolved before releasing the app's first stable release. See [help.nextcloud.com](https://help.nextcloud.com/t/beta-needs-testing-pico-cms-for-nextcloud-v1-0-0/58718) for more info.

This should be fixed in Pico CMS for Nextcloud v1.0.0. You can check it out now to test whether your issue persists or not. If it does, please open a new Issue here on GitHub.
